File: backend/app/schemas/task.py
# backend/app/schemas/task.py
from datetime import date, datetime
from typing import Literal, Optional

from pydantic import BaseModel, Field

Priority = Literal["urgente", "alta", "media", "baja"]
TaskStatus = Literal["pendiente", "en_progreso", "completada"]


class ChecklistItemRead(BaseModel):
    id: int
    task_id: int
    text: str
    is_done: bool
    sort_order: int
    model_config = {"from_attributes": True}


class ChecklistItemCreate(BaseModel):
    text: str = Field(min_length=1, max_length=500)
    sort_order: int = 0


class ChecklistItemUpdate(BaseModel):
    text: Optional[str] = Field(default=None, max_length=500)
    is_done: Optional[bool] = None
    sort_order: Optional[int] = None


class TaskCreate(BaseModel):
    title: str = Field(min_length=1, max_length=300)
    description: Optional[str] = None
    priority: Priority = "media"
    due_date: Optional[date] = None
    sort_order: int = 0
    checklist_items: list[ChecklistItemCreate] = Field(default_factory=list)


class TaskUpdate(BaseModel):
    title: Optional[str] = Field(default=None, max_length=300)
    description: Optional[str] = None
    priority: Optional[Priority] = None
    due_date: Optional[date] = None
    status: Optional[TaskStatus] = None
    sort_order: Optional[int] = None


class TaskRead(BaseModel):
    id: int
    project_id: int
    created_by: int
    title: str
    description: Optional[str]
    priority: str
    status: str
    due_date: Optional[date]
    sort_order: int
    evidence_hint: str
    completed_at: Optional[datetime]
    completed_by: Optional[int]
    created_at: datetime
    updated_at: datetime
    checklist_items: list[ChecklistItemRead] = Field(default_factory=list)
    checklist_total: int = 0
    checklist_done: int = 0
    model_config = {"from_attributes": True}


class ProjectProgress(BaseModel):
    project_id: int
    total_tasks: int
    completed_tasks: int
    progress_pct: float

File: frontend/src/types/task.ts
export type TaskPriority = "urgente" | "alta" | "media" | "baja"
export type TaskStatus = "pendiente" | "en_progreso" | "completada" | "cancelada"

export interface ChecklistItem {
  id: number
  task_id: number
  text: string
  is_done: boolean
  sort_order: number
}

export interface ProjectTask {
  id: number
  project_id: number
  created_by: number
  title: string
  description: string | null
  priority: TaskPriority
  status: TaskStatus
  due_date: string | null
  sort_order: number
  evidence_hint: string
  completed_at: string | null
  completed_by: number | null
  created_at: string
  updated_at: string
  checklist_items: ChecklistItem[]
  checklist_total: number
  checklist_done: number
}

export interface ProjectProgress {
  project_id: number
  total_tasks: number
  completed_tasks: number
  progress_pct: number
}

export interface TaskCreate {
  title: string
  description?: string
  priority?: TaskPriority
  due_date?: string
  sort_order?: number
  checklist_items?: { text: string; sort_order?: number }[]
}

export interface TaskUpdate {
  title?: string
  description?: string
  priority?: TaskPriority
  due_date?: string | null
  status?: TaskStatus
  sort_order?: number
}
